core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/AbstractMetadataInputFormat.java

```diff
@@ -0,0 +1,171 @@
+import java.util.PriorityQueue;
+
+
+/**
+ * Abstract class that implements the inputFormat for the FileCopySource plugin to
+ * read file metadata. The getsplit method creates splits according to user-set configuration
+ * and tries to assign files to each split such that every split copies roughly the same number of bytes.
+ * @param <KEY>
+ * @param <VALUE>
+ */
+public abstract class AbstractMetadataInputFormat<KEY, VALUE> extends InputFormat<KEY, VALUE> {
+
+  protected static final String SOURCE_PATHS = "source.paths";
+  protected static final String MAX_SPLIT_SIZE = "max.split.size";
+  protected static final String FS_URI = "uri";
+  protected static final String RECURSIVE_COPY = "recursive.copy";
+  protected static final int DEFAULT_MAX_SPLIT_SIZE = 128;
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractMetadataInputFormat.class);
+
+  public AbstractMetadataInputFormat() {
+  }
+
+  @Override
+  public List<InputSplit> getSplits(JobContext jobContext) throws IOException, InterruptedException {
+    // connect to the source filesystem
+    Configuration conf = jobContext.getConfiguration();
+    URI uri = URI.create(conf.get(FS_URI));
+    String[] sourcePaths = conf.get(SOURCE_PATHS).split(",");
+    Boolean recursiveCopy = conf.getBoolean(RECURSIVE_COPY, true);
+    FileSystem fileSystem = FileSystem.get(uri, conf);
+    int maxSplitSize = conf.getInt(MAX_SPLIT_SIZE, DEFAULT_MAX_SPLIT_SIZE);
+
+    List<FileStatus> fileStatuses = new ArrayList<>();
+    List<String> baseFolders = new ArrayList<>();
+    for (String prefix : sourcePaths) {
+      try {
+        FileStatus fileStatus = fileSystem.getFileStatus(new Path(prefix));
+        if (fileStatus.isFile()) {
+          prefix = "";
+        }
+        recursivelyAddFileStatus(fileStatuses, baseFolders, prefix, fileStatus, recursiveCopy, fileSystem);
+      } catch (Exception e) {
+        LOG.warn(e.getMessage());
+      }
+    }
+
+    fileSystem.close();
+
+    // create a filemetadata list of all the files
+    List<AbstractFileMetadata> fileMetaDataList = getFileMetadaDataList(fileStatuses, baseFolders,
+                                                                        fileStatuses.size(),
+                                                                        getCredentialsFromConf(conf));
+
+    // compute number of splits and instantiate the splits
+    int numSplits = (fileMetaDataList.size() - 1) / maxSplitSize + 1;
+    PriorityQueue<AbstractMetadataInputSplit> abstractInputSplits = new PriorityQueue<>(numSplits);
+    for (int i = 0; i < numSplits; i++) {
+      abstractInputSplits.add(getInputSplit());
+    }
+
+    // assign each split approximately the same number of bytes (2-approx)
+    for (int i = 0; i < fileMetaDataList.size(); i++) {
+      AbstractMetadataInputSplit minInputSplit = abstractInputSplits.poll();
+      minInputSplit.addFileMetadata(fileMetaDataList.get(i));
+      abstractInputSplits.add(minInputSplit);
```

Comment:
Yeah that's one of the concerns I had when I wrote this method. Consider mapping the following list of files to only 2 splits.
one 100GB file ....... 1000 50B files ......... one 100GB file.
If we enforce the number of files in maxSplitSize, the two big files would end up in the same split.

I agree that the pipeline should behave how the users expect it to behave. Perhaps sorting the files by size first would be a better idea?